mops/src/opsa/opsa.cu

```diff
@@ -1 +1,393 @@
+                }
+
+                // thread 0 contains the gradient for this subset of features_A.
+                if (threadCol == 0) {
+                    buffer_grad_A[i * nThreadRow + threadRow] = dsumA;
+                }
+            }
+        }
+
+        __syncwarp();
+
+        if (grad_B != nullptr) {
+            // write gradB
+            for (int j = threadCol; j < nfeatures_B; j += WARP_SIZE) {
+                grad_B[edge * nfeatures_B + j] = buffer_grad_B[threadRow * nfeatures_B + j];
+            }
+        }
+
+        if (grad_A != nullptr) {
+            // write gradA
+            for (int i = threadCol; i < nfeatures_A; i += WARP_SIZE) {
+                grad_A[edge * nfeatures_A + i] = buffer_grad_A[i * nThreadRow + threadRow];
+            }
+        }
+    }
+}
+
+template <typename scalar_t>
+void mops::cuda::outer_product_scatter_add_vjp(
+    Tensor<scalar_t, 2> grad_A,
+    Tensor<scalar_t, 2> grad_B,
+    Tensor<scalar_t, 3> grad_output,
+    Tensor<scalar_t, 2> A,
+    Tensor<scalar_t, 2> B,
+    Tensor<int32_t, 1> indices_output
+) {
+
+    check_sizes(A, "A", 0, B, "B", 0, "cuda_opsa_vjp");
+    check_sizes(A, "A", 1, grad_output, "grad_output", 1, "cuda_opsa_vjp");
+    check_sizes(B, "B", 1, grad_output, "grad_output", 2, "cuda_opsa_vjp");
+    check_sizes(A, "A", 0, indices_output, "indices_output", 0, "cuda_opsa_vjp");
+
+    int32_t nedges = A.shape[0];
+    int32_t nnodes = grad_output.shape[0];
+    int32_t nfeatures_A = grad_output.shape[1];
+    int32_t nfeatures_B = grad_output.shape[2];
+
+    int32_t *first_occurences = calculate_first_occurences_cuda(indices_output.data, nedges, nnodes);
+
+    dim3 gridDim(nnodes, 1, 1);
+
+    dim3 blockDim(NWARPS_PER_BLOCK * WARP_SIZE, 1, 1);
+
+    void *sptr = 0;
+    size_t space = 0;
+
+    shared_array<scalar_t>(nfeatures_A * nfeatures_B, sptr, &space);
+    shared_array<scalar_t>(NWARPS_PER_BLOCK * nfeatures_A, sptr, &space);
+    shared_array<scalar_t>(NWARPS_PER_BLOCK * nfeatures_B, sptr, &space);
+
+    if (grad_A.data != nullptr) {
+        check_same_shape(grad_A, "grad_A", A, "A", "cuda_opsa_vjp");
+        shared_array<scalar_t>(NWARPS_PER_BLOCK * nfeatures_A, sptr, &space);
+    }
+    if (grad_B.data != nullptr) {
+        check_same_shape(grad_B, "grad_B", B, "B", "cuda_opsa_vjp");
+        shared_array<scalar_t>(NWARPS_PER_BLOCK * nfeatures_B, sptr, &space);
+    }
+
+    outer_product_scatter_add_vjp_kernel<scalar_t><<<gridDim, blockDim, space>>>(
+        A.data,
+        B.data,
+        nnodes,
+        nedges,
+        nfeatures_A,
+        nfeatures_B,
+        first_occurences,
+        indices_output.data,
+        grad_output.data,
+        grad_A.data,
+        grad_B.data
+    );
+
+    CUDA_CHECK_ERROR(cudaGetLastError());
+
+    CUDA_CHECK_ERROR(cudaDeviceSynchronize());
+}
+
+// these templates will be precompiled and provided in the mops library
```

Comment:
comment looks wrong

Reply:
removed comment - I copied it from the CPU code.